torch_xla/csrc/runtime/tensor_source.h

```diff
@@ -53,10 +53,9 @@ class AtenSource : public TensorSource {
     at::ScalarType target_torch_type = TorchTypeFromXlaType(primitive_type());
     if (target_torch_type != tensor.type().scalarType()) {
       TORCH_LAZY_COUNTER("AtenSourceDowncasts", 1);
-      tensor_ = std::move(tensor.to(target_torch_type).contiguous());
-    } else {
-      tensor_ = std::move(tensor.contiguous());
     }
+    tensor_ = std::move(tensor.to(target_torch_type, /*non_blocking=*/false,
+                                  /*copy=*/true, at::MemoryFormat::Contiguous));
```

Comment:
thinking.. we have 2 options, either copy the tensor on CPU or only return the control to python after we started the transfer(or finish the transfer? hard for me to tell at which stage the origional tensor is not needed).

If I understand correctly instead of creating a XLA:Literal, now we perform a copy on the cpu tensor, and copying the cpu tensor is faster?

Reply:
> thinking.. we have 2 options, either copy the tensor on CPU or only return the control to python after we started the transfer(or finish the transfer? hard for me to tell at which stage the origional tensor is not needed).

Correct. We don't have any tools to "lock" the input tensor, so we should either block until the transfer is done or we make a CPU copy here (which is also blocking, but faster than the CPU -> TPU copy).

The third option is to let the caller decide via the `non_blocking` argument. Whether the CPU tensor will ever be modified during transfer is context dependent, so the caller can decide whether an unsafe concurrent copy is okay. We very likely want to set `non_blocking=True` in our data loader, for example. The default case (`non_blocking=False`) would skip the copy if the tensor is already contiguous and has the correct dtype, saving host memory. This make the default case slower, but it also makes it safer (avoiding OOMs and races) and more consistent with upstream/eager (where `.to` is blocking by default).

> If I understand correctly instead of creating a XLA:Literal, now we perform a copy on the cpu tensor, and copying the cpu tensor is faster?

Yeah. I couldn't tell you why this is faster, but it is.